### PR TITLE
Add more DT/EM/section constants

### PR DESCRIPTION
**Pull request by Bo98**
These are largely from binutils.

### CODE REVIEW

**Comment by david942j** [comment]
> > ```
> > lib/elftools/constants.rb:54:5: C: Metrics/ModuleLength: Module has too many lines. [122/100]
> >    module DT ...
> >    ^^^^^^^^^
> > ```
> 
> I don't think I can really fix that given it's an enumeration module, so I've added the file as an exclusion if that's ok.

sounds good to me

File: lib/elftools/constants.rb [hunk 1]
```diff
@@ -42,120 +42,371 @@ module DF
...
+      DT_TEXTREL                    = 22 # has been superseded by the DF_TEXTREL flag
+      DT_JMPREL                     = 23 # address of relocation entries associated solely with procedure linkage table
+      DT_BIND_NOW                   = 24 # if the loader needs to do relocate now, superseded by the DF_BIND_NOW flag
+      DT_INIT_ARRAY                 = 25 # address init array
+      DT_FINI_ARRAY                 = 26 # address of fini array
+      DT_INIT_ARRAYSZ               = 27 # total size of init array
+      DT_FINI_ARRAYSZ               = 28 # total size of fini array
+      DT_RUNPATH                    = 29 # path of libraries for searching
+      DT_FLAGS                      = 30 # flags
+      DT_ENCODING                   = 32 # just a lower bound
```
Comment (david942j) [inline lib/elftools/constants.rb:86]: Seems can be removed as it has the same value as DT_PREINIT_ARRAY. Is it still used in binutils?

File: lib/elftools/constants.rb [hunk 2]
```diff
@@ -42,120 +42,371 @@ module DF
...
+      DT_TLSDESC_GOT                = 0x6ffffef7 # :nodoc:
+      DT_GNU_CONFLICT               = 0x6ffffef8 # :nodoc:
+      DT_GNU_LIBLIST                = 0x6ffffef9 # :nodoc:
+      DT_CONFIG                     = 0x6ffffefa # :nodoc:
+      DT_DEPAUDIT                   = 0x6ffffefb # :nodoc:
+      DT_AUDIT                      = 0x6ffffefc # :nodoc:
+      DT_PLTPAD                     = 0x6ffffefd # :nodoc:
+      DT_MOVETAB                    = 0x6ffffefe # :nodoc:
+      DT_SYMINFO                    = 0x6ffffeff # :nodoc:
+      DT_ADDRRNGHI                  = 0x6ffffeff # see {DT_ADDRRNGLO}
```
Comment (david942j) [inline lib/elftools/constants.rb:115]: same here, is it still used?

File: lib/elftools/constants.rb (hunk 1 above)
Comment (Bo98) [reply lib/elftools/constants.rb:86]: `DT_ENCODING` isn't a real enum entry but a special value:

> A tag whose value is an even number indicates a dynamic section entry that uses `d_ptr`. A tag whose value is an odd number indicates a dynamic section entry that uses `d_val` or that uses neither `d_ptr` nor `d_val`. Tags whose values are less than the special value `DT_ENCODING` and tags whose values fall between `DT_HIOS` and `DT_LOPROC` do not follow these rules. 

Looking at the the values after:

```
DT_PREINIT_ARRAY -> even -> d_ptr -> correct given it's an address
DT_PREINIT_ARRAYSZ -> odd -> d_val -> correct given it's a size
DT_SYMTAB_SHNDX -> even -> d_ptr -> correct given it's an address
DT_RELRSZ -> odd -> d_val -> correct given it's a size
DT_RELR -> even -> d_ptr -> correct given it's an address
DT_RELRENT -> odd -> d_val -> correct given it's a count
```

When `DT_RELRENT` was added, `DT_ENCODING` did get bumped in bintuils. This change is however disputed (it hasn't changed anywhere else like LLVM, glibc or even elfcpp from binutils) and I do not believe the change was correct.

File: lib/elftools/constants.rb (hunk 2 above)
Comment (Bo98) [reply lib/elftools/constants.rb:115]: Similar story to `DT_ENCODING`. `DT_ADDRRNGLO` to `DT_ADDRRNGHI` specify a range of values which use the `d_ptr`. This range is inclusive so overlaps with `DT_SYMINFO`.

File: lib/elftools/constants.rb (hunk 1 above)
Comment (david942j) [reply lib/elftools/constants.rb:86]: Thanks for the detailed explanation!

**Comment by david942j** [comment]
Merged. Thanks for the contribution!

**Pull request merged**